pkg/kube/krt/nestedjoin.go

```diff
@@ -0,0 +1,282 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"sync"
+
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/util/sets"
+)
+
+type nestedjoin[T any] struct {
+	collectionName           string
+	id                       collectionUID
+	collections              internalCollection[Collection[T]]
+	synced                   <-chan struct{}
+	uncheckedOverlap         bool
+	syncer                   Syncer
+	collectionChangeHandlers []func(collectionChangeEvent[T])
+	sync.RWMutex
+}
+
+func (j *nestedjoin[T]) GetKey(k string) *T {
+	for _, c := range j.collections.List() {
+		if r := c.GetKey(k); r != nil {
+			return r
+		}
+	}
+	return nil
+}
```

Comment:
A musing about the concept of "first".

With normal join, first is probably mostly static. You have a bunch of "ServiceInfo" collections for instance, and then you join them from `[]Collection[ServiceInfo]{fromServices, fromServiceEntries}` so if there's key conflict "first" should come out of `fromService`.

With nestedJoin we don't have the same static membership, it is the raison d'etre. It seems to me this could lead to an undefined "first".

Reply:
*Technically applies to List as well

**I don't have any solution to offer, I'm not convinced this even a problem krt must solve vs a usage concern.

Reply:
This is exactly correct; first is not stable. What i'm unclear on is how necessary/desirable it is to keep a stable first. If membership changes, then maybe you do want first to change 🤷🏾‍♂️ regardless, the follow-up PR that builds on top of this implements merging so that callers can get deterministic output

Reply:
That's fair. I do think the merge follow helps, it provides better tools to deal with this at the very least.

Reply:
What does "stable ordering" mean specifically? Is it just a sort func? Or some way of ensuring that collections show up in the sameish index?

Reply:
Lets say my merge function is "take the first one". Its going to flip-flop every time we build this (List is unordered).

A more practical example is merging model.Service in istio: https://github.com/istio/istio/blob/3404613f57cb5447ccb779b97c1e39ef79332ae8/pilot/pkg/serviceregistry/aggregate/controller.go#L327. _Order matters_ here (for any field except port/IP)

Reply:
Ack, so we should be able to just sort by collection uids then; that way we don't flip flop

Reply:
I think we probably have the same issue in current MC code but that only guarantees consistency within a process, each Istiod instance would (potentially) have a different ordering.

I think it would be very good to have non-order-sensitive merging though ideally. The service case is pretty broken.

One semi-related note: the merge _probably_ wants to treat "local cluster" vs "remote cluster" differently. I am not sure we can see that?

Reply:
> I think it would be very good to have non-order-sensitive merging though ideally. The service case is pretty broken.

Agree with this; I'm planning to write my own merging functions that don't assume ordering

> One semi-related note: the merge probably wants to treat "local cluster" vs "remote cluster" differently. I am not sure we can see that?

I just ran into this while implementing the controller portions; I wonder if we can get around this by having the global collections call Fetch(LocalCluster) in the merge function. If not, there could be a "priority nested join" or something like that

Reply:
> I think we probably have the same issue in current MC code but that only guarantees consistency within a process, each Istiod instance would (potentially) have a different ordering.

If the merge functions don't assume ordering, is this still an issue? That being said, we're at least as good as current multi-cluster, so that makes me feel a bit better about the current state

Reply:
> I just ran into this while implementing the controller portions; I wonder if we can get around this by having the global collections call Fetch(LocalCluster) in the merge function. If not, there could be a "priority nested join" or something like that

No handler context 😢 I threw something together where a client can pass a "primary" merge function where the primary T is passed to the func. Not sure how generally useful it is outside of multi-cluster, but a new field felt lighter than a whole new type

Reply:
We discussed this 1:1 - we can add decorators to the objects themselves that allow us to determine the local vs. remote cluster